### PR TITLE
[IMP][UAASDEV-7371] `account_financial_report_webkit` - Improving partner balance report

### PR DESCRIPTION
**Pull request by pnajman-modoolar**
[IMP][UAASDEV-7371] `account_financial_report_webkit` - Improving partner balance report to work properly in combination with `account.static.balance` by excluding std Odoo opening periods.

### CODE REVIEW

File: account_financial_report_webkit/report/common_partner_reports.py [hunk 1]
```diff
@@ -89,17 +89,20 @@ def _get_first_special_period(self):
     def _get_period_range_from_start_period(self, start_period,
                                             include_opening=False,
                                             fiscalyear=False,
-                                            stop_at_previous_opening=False):
+                                            stop_at_previous_opening=False,
+                                            **kwargs):
```
Comment (PieterPaulussen) [inline account_financial_report_webkit/report/common_partner_reports.py:93]: Is there a specific reason you used the `kwargs` key instead of a named param, or what is the benefit of being able to pass more parameters to this method?

Comment (pnajman-modoolar) [reply account_financial_report_webkit/report/common_partner_reports.py:93]: yes, this is override of [this method](https://github.com/EssentNovaTeam/account-financial-reporting/pull/3/files#diff-07dff3641d9dda2dfeb17d89bcd41ba4L314). I needed to find a way to optionally exclude `first special period` without extending base method with the specific named parameter which  is needed here as it does not make sense to do something like that, so the only other option was to go with ``kwargs``.

Comment (StefanRijnhart) [reply account_financial_report_webkit/report/common_partner_reports.py:93]: Is this something that can be added to the context instead?

Comment (pnajman-modoolar) [reply account_financial_report_webkit/report/common_partner_reports.py:93]: That was my first approach but unfortunately these classes don't have context attribute.

Comment (pnajman-modoolar) [reply account_financial_report_webkit/report/common_partner_reports.py:93]: It's painful, I know, but from my perspective this was the only way to go in order to keep changes on minimum.

Comment (pnajman-modoolar) [reply account_financial_report_webkit/report/common_partner_reports.py:93]: Further more, context/localcontext comes with class `report_sxw.rml_parse` and that class is extended by the actual report classes (i.e. `PartnerBalance`, `PartnersLedger`, etc) but the classes I extended here do not have such an attribute.

**Review by PieterPaulussen — APPROVED** [review]

File: account_financial_report_webkit/report/common_partner_reports.py (hunk 1 above)
Comment (PieterPaulussen) [reply account_financial_report_webkit/report/common_partner_reports.py:93]: Thanks for the clarification!

**Review by PieterPaulussen — APPROVED** [review]

**Review by StefanRijnhart — APPROVED** [review]

**Comment by pnajman-modoolar** [comment]
@StefanRijnhart @PieterPaulussen, guys,

Just letting you know that tests are still failing [here](https://bitbucket.org/emagine-reality/essentbe-odoo/pull-requests/1670/imp-uaasdev-7371-partner-balance-report/diff).

I'll add some debug lines to help me figure out what's going on here ....

File: account_financial_report_webkit/report/common_partner_reports.py [hunk 2]
```diff
@@ -89,17 +89,20 @@ def _get_first_special_period(self):
...
             _get_period_range_from_start_period(
                 start_period,
                 include_opening=include_opening,
                 fiscalyear=fiscalyear,
                 stop_at_previous_opening=stop_at_previous_opening)
-        first_special = self._get_first_special_period()
-        if first_special and first_special.id not in periods:
-            periods.append(first_special.id)
+
+        if not kwargs.get('exclude_first_special_period', False):
```
Comment (StefanRijnhart) [inline account_financial_report_webkit/report/common_partner_reports.py:102]: When you exclude the first special period, this leads to _compute_init_balance being called with an empty list of periods which raises the exception in the tests. Isn't it safe to include the first special period, or rather: will the reports actually be correct without it?

Comment (pnajman-modoolar) [reply account_financial_report_webkit/report/common_partner_reports.py:102]: Yes, in context of `Partner Balance` report (and other reports) we need to exclude first special period as those numbers are gathered via `account.static.balance` `get_balances` method, so the initial balances gets duplicated and thus we have invalid report.

If I'm right here then we have problem with other reports as well not just with partner balance.
So, I added those to log lines to be sure that the context values from yml test file are correctly propagated to the report, because something tells me that those values are missing for some reason.

Let's see ....

Comment (pnajman-modoolar) [reply account_financial_report_webkit/report/common_partner_reports.py:102]: As i was afraid of, context values are not passed from yml test file to the report.
Full Log is [here](https://build.emadev.nl/blue/organizations/jenkins/BillingHouse%2FEssent%2FEssent-BE-Odoo-Jenkinsfile/detail/PR-1670/6/pipeline)

Below are debug lines from the Partner Balance report:
> [flow[60860166-b33f-44e1-b21b-ded430375b0d][test:set_metadata][Administrator][odoo]]2020-01-23 09:39:50,868 UTC 2127 INFO odoo_1 openerp.addons.account_financial_report_webkit.report.partner_balance: include_opening flag set to: 'value not present in context'
[flow[60860166-b33f-44e1-b21b-ded430375b0d][test:set_metadata][Administrator][odoo]]2020-01-23 09:39:50,868 UTC 2127 INFO odoo_1 openerp.addons.account_financial_report_webkit.report.partner_balance: `exclude_first_special_period` flag set to: 'value not present in context'
[flow[60860166-b33f-44e1-b21b-ded430375b0d][test:set_metadata][Administrator][odoo]]2020-01-23 09:39:50,869 UTC 2127 INFO odoo_1 openerp.addons.account_financial_report_webkit.report.partner_balance: `exclude_first_special_period` flag set to: 'value not present in context'

@StefanRijnhart  Do you have any idea why context from yml test is not propagated to the report?

**Comment by pnajman-modoolar** [comment]
It turns out this module works perfectly fine and that the problem is in the data source.
First special period is taken from the fiscal year of the oldest move line in database.
In our database the oldest move lines are the one belonging to the moves below.
Those move lines belongs to fiscal year 2019 but their dates are, well, you can see for yourself.

odoo=> select id, date from account_move where date < '1999-01-01' order by id asc;;
    id    |    date
----------+------------
 43794378 | 0219-04-30
 43796214 | 0219-04-30
(2 rows)\

Bottom line is, we need to fix these moves on dev/uat environments as well as on production and also to merge this [PR](https://bitbucket.org/emagine-reality/essentbe-odoo/pull-requests/1671/imp-uaasdev-7371/diff).

I'll close this PR as it's not longer needed
